src/charm.py

```diff
@@ -4,27 +4,289 @@
+            )
+
+    def _on_leader_elected(self, _):
+        """Handle the leader elected event.
+
+        Set config values in the peer relation databag.
+        """
+        peer_data = self._peers.data[self.app]
+
+        required_passwords = ["root-password", "server-config-password", "cluster-admin-password"]
+
+        for required_password in required_passwords:
+            if not peer_data.get(required_password):
+                logger.debug(f"Setting {required_password}")
+                password = generate_random_password(PASSWORD_LENGTH)
+                peer_data[required_password] = password
+
+    def _on_mysql_pebble_ready(self, event):
+        """Pebble ready handler.
+
+        Define and start a pebble service and bootstrap instance.
+        """
+        if not self._is_peer_data_set:
+            self.unit.status = WaitingStatus("Waiting for leader election.")
+            logger.debug("Leader not ready yet, waiting...")
+            event.defer()
+            return
+
+        container = event.workload
+
+        if not container.exists(CONFIGURED_FILE):
+            # First run setup
+            self.unit.status = MaintenanceStatus("Initialising mysqld")
+            try:
+                # Allow mysql instances to reach each other
+                self._mysql.patch_dns_searches(self.app.name)
+
+                # Run mysqld for the first time to
+                # bootstrap the data directory and users
+                logger.debug("Initialising instance")
+                self._mysql.initialise_mysqld()
+
+                # Add the pebble layer
+                container.add_layer(MYSQLD_SERVICE, self._pebble_layer, combine=False)
+                container.restart(MYSQLD_SERVICE)
+                logger.debug("Waiting for instance to be ready")
+                self._mysql._wait_until_mysql_connection()
+
+                logger.info("Configuring instance")
+                # Configure all base users and revoke
+                # privileges from the root users
+                self._mysql.configure_mysql_users()
+                # Configure instance as a cluster node
+                self._mysql.configure_instance()
+
+            except (
+                MySQLConfigureInstanceError,
+                MySQLConfigureMySQLUsersError,
+                MySQLInitialiseMySQLDError,
+                MySQLPatchDNSSearchesError,
+            ) as e:
+                self.unit.status = BlockedStatus("Unable to configure instance")
+                logger.debug("Unable to configure instance: {}".format(e))
+                return
+
+            if self.unit.is_leader():
+                try:
+                    # Create the cluster when is the leader unit
+                    self._mysql.create_cluster()
+                except MySQLCreateClusterError as e:
+                    self.unit.status = BlockedStatus("Unable to create cluster")
```

Comment:
Cant disagree, but bear with me. We have two options here, which are:
1. Not treat the exception and allow the charm to enter in a error state
2. Treat the exception (as implemented) and leave it blocked.

Both indicates a failures and requires user intervention. So the decision boils up to how the treatment will be done.
Canonically, a blockedstatus should be resolvable through an action, but I'm not completely sure though.
For now, the current approach is to prevent ugly non treated exceptions reach the end user.

Reply:
Is there no way to declare a Failed status from within a charm?

Reply:
Not really. From statuses declaration, `BlockedStatus` is the most severe one.